### DISCUSSION TITLE
Pluggable GraphQL libraries

### DISCUSSION THREAD

**Discussion by zerolab** [torchbox/wagtail-grapple#163] [Ideas]
Given the [recent developments](https://github.com/graphql-python/graphene/issues/1312) with Graphene, one idea would be to make the GraphQL libraries swappable (e.g. [Strawberry](https://strawberry.rocks/) is the new hot topic in this area).

This would require a fair bit of plumbing, but it is an enticing thought. Does anyone have examples of packages that do this well and are worth following?

**Comment by fabienheureux** [comment]
I think that API Platform might be interesting here https://github.com/api-platform/api-platform

They provide various formats for the data model.
We could provide, through Grapple, a way to translate Django models to GraphQL schemas OR some other well-supported formats (JSON-LD, Hydra...)

This is what ariadne-django partially achieve https://github.com/reset-button/ariadne_django (current supported package to bring the gap between ariadne and django)

**Comment by timmysmalls** [comment]
I'm all for it, but it does sound like a pretty big task. I guess it all depends on how easy it is to find a common denominator between the different Django GQL packages, and what kind of features we want to add. I can see something like that holding development back in the future if we want to add features that are easy to implement in one library, but more difficult in another. Also, do we want to make Grapple completely back-end agnostic, so that anyone could write a new GQL package and plug it in, or do we just want to support a few existing libraries? The latter is probably easier, but it does mean that we'd be expected to maintain the integration with the different GQL libraries.
